src/bms/parse/random/ast_parse.rs

```diff
@@ -0,0 +1,277 @@
+use crate::bms::{lex::token::Token, parse::rng::Rng};
+
+use super::ControlFlowRule;
+use super::ast_build::*;
+
+pub(super) fn parse_control_flow_ast<'a>(
+    iter: &mut std::iter::Peekable<impl Iterator<Item = Unit<'a>>>,
+    rng: &mut impl Rng,
+    error_list: &mut Vec<ControlFlowRule>,
+) -> Vec<&'a Token<'a>> {
+    let mut result = Vec::new();
+    while let Some(unit) = iter.next() {
+        match unit {
+            Unit::Token(token) => {
+                result.push(token);
+            }
+            Unit::RandomBlock { value, if_blocks } => {
+                // Select branch
+                let branch_val = match value {
+                    BlockValue::Random { max } => {
+                        if max == 0 {
+                            0
+                        } else {
+                            rng.generate(1..=(max as u32)) as u64
+                        }
+                    }
+                    BlockValue::Set { value } => value,
+                };
+                // Find the branch in the first if_block that contains this branch value
+                let mut found = false;
+                for if_block in &if_blocks {
+                    if let Some(branch) = if_block.branches.get(&branch_val) {
+                        let mut branch_iter = branch.tokens.clone().into_iter().peekable();
+                        result.extend(parse_control_flow_ast(&mut branch_iter, rng, error_list));
+                        found = true;
+                        break;
+                    }
+                }
+                // If not found, try to find the 0 (else) branch
+                if !found {
+                    for if_block in &if_blocks {
+                        #[allow(unused_assignments)]
+                        if let Some(branch) = if_block.branches.get(&0) {
+                            let mut branch_iter = branch.tokens.clone().into_iter().peekable();
+                            result.extend(parse_control_flow_ast(
+                                &mut branch_iter,
+                                rng,
+                                error_list,
+                            ));
+                            found = true;
+                            break;
+                        }
+                    }
```

Comment:
You can write this more simply:
```suggestion
                    
                    if let Some(else_branch) = if_blocks
                        .iter()
                        .flat_map(|if_block| if_block.branches.get(&0))
                        .next()
                    {
                        let mut branch_iter = else_branch.tokens.clone().into_iter().peekable();
                        result.extend(parse_control_flow_ast(&mut branch_iter, rng, error_list));
                    }
```